src/Runtime/jni/jniwrapper.c

```diff
@@ -910,3 +909,23 @@ JNIEXPORT jstring JNICALL Java_com_ibm_onnxmlir_OMModel_output_1signature_1jni(
 
   return java_osig;
 }
+
+JNIEXPORT jobject JNICALL Java_com_ibm_onnxmlir_OMTensor_free_1data_1jni(
+    JNIEnv *env, jclass cls, jobject java_data) {
+
+  log_init();
+
+  /* Find and initialize Java Exception class */
+  JNI_TYPE_VAR_CALL(env, jclass, jecpt_cls,
+      (*env)->FindClass(env, jnistr[CLS_JAVA_LANG_EXCEPTION]),
+      jecpt_cls != NULL, NULL, "Class java/lang/Exception not found");
+
+  /* Get native data buffer backing direct byte buffer */
+  JNI_TYPE_VAR_CALL(env, void *, jni_data,
+      (*env)->GetDirectBufferAddress(env, java_data), jni_data != NULL,
+      jecpt_cls, "jni_data=%p", jni_data);
+
+  free(jni_data);
```

Comment:
I ran into this during testing of my older `finalize()`-based PR, but `jni_data` points to the aligned allocation, not to the start of the allocation.

Reply:
I see what you are saying but I'm a bit surprised why this would happen. Basically, when you call `NewDirectByteBuffer`, you give it the pointer to the memory block. Now when you call `GetDirectBufferAddress`, it's going to give you a different (aligned) pointer? I mean I gave it the pointer, now I want it back, why would it change it behind my back?!

Reply:
OK maybe you mean onnx-mlir might give you an aligned pointer that's not the start of the allocation? OK in that case we need fiddle a bit like the offset thing you did. Itchy.

Reply:
OK I looked at the conversion on your finalize PR and you said:

>I thought about doing that but I wanted to avoid a limitation with java - all of the integral based arguments are signed. With an offset, which can be signed, I can keep the semantics more or less the same. With a Cleaner based approach I can create another DirectByteBuffer holding the actual allocation variable which the cleaner can use to clean up - no offset calculation required.

So I think instead of passing in the offset, we could just pass in the aligned pointer as jlong, right? As you said later, it doesn't really matter if it's signed or not. 64-bit is 64-bit. It only matters how you interpret it. Doing that would save us from having to keep another flag.